sql/src/main/java/io/crate/analyze/TableIdentsExtractor.java

```diff
@@ -0,0 +1,218 @@
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze;
+
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.analyze.relations.AnalyzedRelationVisitor;
+import io.crate.analyze.relations.DocTableRelation;
+import io.crate.analyze.relations.QueriedDocTable;
+import io.crate.analyze.relations.TableFunctionRelation;
+import io.crate.analyze.relations.TableRelation;
+import io.crate.analyze.symbol.Aggregation;
+import io.crate.analyze.symbol.DynamicReference;
+import io.crate.analyze.symbol.FetchReference;
+import io.crate.analyze.symbol.Field;
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.InputColumn;
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.MatchPredicate;
+import io.crate.analyze.symbol.ParameterSymbol;
+import io.crate.analyze.symbol.SelectSymbol;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.analyze.symbol.SymbolVisitor;
+import io.crate.analyze.symbol.Value;
+import io.crate.metadata.Reference;
+import io.crate.metadata.TableIdent;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+public class TableIdentsExtractor {
+
+    private static final TableIdentRelationVisitor RELATION_VISITOR = new TableIdentRelationVisitor();
+    private static final TableIdentSymbolVisitor SYMBOL_VISITOR = new TableIdentSymbolVisitor();
+
+    /**
+     * Extracts all table idents from all given symbols if possible (some symbols don't provide any table ident info)
+     */
+    public static Collection<TableIdent> extract(Iterable<? extends Symbol> symbols) {
+        Collection<TableIdent> tableIdents = null;
+        for (Symbol symbol : symbols) {
+            Collection<TableIdent> tableIdentList = SYMBOL_VISITOR.process(symbol, null);
+            //noinspection PointlessBooleanExpression
+            if (tableIdentList != null && tableIdentList.isEmpty() == false) {
+                if (tableIdents == null) {
+                    tableIdents = new HashSet<>(tableIdentList.size());
```

Comment:
not sure if this attempted optimization is really improving things - it might break method inlining and mess up branch prediction